pkg/virtualKubelet/forge/pods.go

```diff
@@ -346,18 +359,23 @@ func RemoteVolumes(volumes []corev1.Volume, enableAPIServerSupport bool, saSecre
 					// Replace the certification authority configmap with the remapped name.
 					source.ConfigMap.Name = RemoteConfigMapName(source.ConfigMap.Name)
 				} else if source.ServiceAccountToken != nil {
-					if !enableAPIServerSupport {
+					if apiServerSupport == APIServerSupportDisabled {
 						// Remove the entry referring to the service account.
 						volumes[i].Projected.Sources = append(volumes[i].Projected.Sources[:j], volumes[i].Projected.Sources[j+1:]...)
 						offset++
 						continue
 					}
 
 					// Replace the ServiceAccountToken entry with the corresponding secret one, only in case it is enabled.
+					secretKey := corev1.ServiceAccountTokenKey
+					if apiServerSupport == APIServerSupportTokenAPI {
+						secretKey = ServiceAccountTokenKey(volumes[i].Name, source.ServiceAccountToken.Path)
+					}
+
 					source.ServiceAccountToken = nil
```

Comment:
this looks like it is dropping audience information from `source.ServiceAccountToken`, which is critical to inject the correct token... without that, this will default to injecting tokens valid for use against the kubernetes api-server, which could enable a privilege escalation if the user had requested a token scoped to a different audience

Reply:
Thanks @liggitt for your comment.

Still, I feel that I might be missing something, as the current behavior seems correct to me. 

Let me give a bit of context. The idea here is to enable pods which are offloaded to remote clusters to contact back the origin API server, as well as support scenarios in which offloaded pods use ServiceAccount tokens with specific audiences to authenticate among each others (e.g., service mesh scenarios).

The solution proposed involves a two step process:

1. A controller recognizes that a new pod mounting one or more service account tokens has been offloaded to a remote cluster. It requests the creation of the token(s) from the origin API server (for the requested audience), and saves them within a secret in the remote cluster. Additionally, it takes care of the refresh/deletion process when necessary. This reflection logic is implemented [here](https://github.com/giorio94/liqo/blob/mio%2Fsa-reflection-token-api/pkg/virtualKubelet/reflection/configuration/serviceaccount.go), and the *TokenRequest* is forged [here](https://github.com/giorio94/liqo/blob/a9b6d49c3b42f5fceba9b0dc457ac6690a65ea9f/pkg/virtualKubelet/forge/secrets.go#L230-L244).
2. The pod specification (and in particular that of the volumes) propagated to the remote cluster is mutated to account for this, changing the ServiceAccountToken references (since the kubelet on the target node would request the tokens from the remote API server) to the Secret entries created by the logic part of step 1 (and containing the tokens valid for the origin cluster). This logic is the one you mentioned, and available [here](https://github.com/giorio94/liqo/blob/a9b6d49c3b42f5fceba9b0dc457ac6690a65ea9f/pkg/virtualKubelet/forge/pods.go#L371-L382).

The `ServiceAccountTokenKey` function you mentioned is not responsible for retrieving the actual service account token (I see that the name might be a bit misleading), but only the key identifying the entry within the secret containing the actual token. This key is the concatenation of the name of the volume itself, and an hash of the path where that token will be mounted (cf. [here](https://github.com/giorio94/liqo/blob/mio%2Fsa-reflection-token-api/pkg/virtualKubelet/forge/secrets.go#L117-L123)). It does not include any audience information (although it would be possible), since two ServiceAccountToken volume entries cannot share the same volume name and path AFAIU. Yet, the token itself is scoped with the requested audience, thus preventing privilege escalation.

Reply:
> This key is the concatenation of the name of the volume itself, and an hash of the path where that token will be mounted (cf. [here](https://github.com/giorio94/liqo/blob/mio%2Fsa-reflection-token-api/pkg/virtualKubelet/forge/secrets.go#L117-L123)). It does not include any audience information (although it would be possible), since two ServiceAccountToken volume entries cannot share the same volume name and path AFAIU

Is the secret unique per pod? If not, two different pods with the same service account, volume name, and mount path could specify two different audiences, which would collide, right?

Reply:
Yes, the secret is unique per pod.

Reply:
> Yes, the secret is unique per pod.

ok, I missed that... tracing the secret name getter looked like it was unique per service account, but not per pod

Reply:
Probably that getter name is a bit misleading since it takes the pod name as parameter...